File: Opentelemetry-Observability-Practice/backend/telemetry.py
"""OpenTelemetry setup for the FastAPI backend.

Exports spans over OTLP/HTTP to Jaeger (default: http://localhost:4318/v1/traces).
"""

from __future__ import annotations

import os

from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor


def setup_telemetry() -> None:
    """
    Configure the global TracerProvider once at process start.

    SERVICE_NAME is how this process shows up in the Jaeger service dropdown.
    """
    # Resource = identity metadata attached to every span from this process.
    resource = Resource.create(
        {
            "service.name": "dashboard-api",
            "service.version": "0.1.0",
        }
    )

    provider = TracerProvider(resource=resource)

    # Where Jaeger listens for OTLP HTTP (docker-compose maps 4318).
    endpoint = os.getenv(
        "OTEL_EXPORTER_OTLP_TRACES_ENDPOINT",
        "http://localhost:4318/v1/traces",
    )

    # BatchSpanProcessor buffers spans and ships them in the background
    # so request handlers stay fast.
    exporter = OTLPSpanExporter(endpoint=endpoint)
    provider.add_span_processor(BatchSpanProcessor(exporter))

    # Register as the process-wide provider used by get_tracer() / instrumentors.
    trace.set_tracer_provider(provider)

File: Opentelemetry-Observability-Practice/frontend/src/telemetry.js
/**
 * Browser OpenTelemetry setup.
 *
 * Creates spans for UI actions and outbound fetch() calls, then exports them
 * to Jaeger via the Vite /otlp proxy. Fetch instrumentation also injects
 * W3C traceparent headers so the FastAPI spans become children of the UI span.
 */

import { ZoneContextManager } from "@opentelemetry/context-zone";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-http";
import { registerInstrumentations } from "@opentelemetry/instrumentation";
import { FetchInstrumentation } from "@opentelemetry/instrumentation-fetch";
import { Resource } from "@opentelemetry/resources";
import { BatchSpanProcessor } from "@opentelemetry/sdk-trace-base";
import { WebTracerProvider } from "@opentelemetry/sdk-trace-web";
import { ATTR_SERVICE_NAME } from "@opentelemetry/semantic-conventions";
import { trace } from "@opentelemetry/api";

let ready = false;

/** Initialize the web tracer once; safe to call from main.jsx before render. */
export function setupTelemetry() {
  if (ready) return;

  // service.name is what appears in Jaeger's service filter for the frontend.
  const resource = new Resource({
    [ATTR_SERVICE_NAME]: "dashboard-ui",
  });

  // Batch + send to Jaeger through the Vite proxy (avoids browser CORS).
  const provider = new WebTracerProvider({ resource });
  provider.addSpanProcessor(
    new BatchSpanProcessor(
      new OTLPTraceExporter({
        url: "/otlp/v1/traces",
      })
    )
  );

  provider.register({
    // Zone.js context manager keeps the active span across async/await.
    contextManager: new ZoneContextManager(),
  });

  registerInstrumentations({
    instrumentations: [
      new FetchInstrumentation({
        // Propagate trace context to our API (and any localhost URL).
        propagateTraceHeaderCorsUrls: [/.*/],
        // Don't create extra spans for the OTLP exporter's own POSTs.
        ignoreUrls: [/\/otlp\//],
      }),
    ],
  });

  ready = true;
}

/** Helper so App.jsx can open named UI spans around button handlers. */
export function getTracer() {
  return trace.getTracer("dashboard-ui");
}
